unicode-data-security/lib/Unicode/Char/Identifiers/Security.hs

```diff
@@ -0,0 +1,115 @@
+-- [TODO] @since
+-- |
+-- Module      : Unicode.Char.Identifiers
+-- Copyright   : (c) 2021 Composewell Technologies and Contributors
+-- License     : Apache-2.0
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+--
+-- Unicode Security Mechanisms functions based on
+-- [Unicode Technical Standard #39](https://www.unicode.org/reports/tr39/)
+
+module Unicode.Char.Identifiers.Security
+    ( -- * Identifier status
+      IdentifierStatus(..)
+    , identifierStatus
+    , isAllowedIdentifier
+
+      -- * Identifier type
+    , T.IdentifierType(..)
+    , identifierTypes
+    , identifierTypeStatus
+
+      -- * Confusables
+    , prototype
+    , prototypeM
+    , intentionalConfusables
+    , isIntentionallyConfusable
+    )
+where
+
+import           Data.List.NonEmpty (NonEmpty)
+import           Data.Maybe (fromMaybe, isJust)
+import qualified GHC.Foreign as Foreign
+import qualified GHC.IO.Encoding as Encoding
+import           System.IO.Unsafe (unsafePerformIO)
+
+import qualified Unicode.Internal.Char.Security.Confusables as C
+import qualified Unicode.Internal.Char.Security.IdentifierStatus as S
+import qualified Unicode.Internal.Char.Security.IdentifierType as T
+import qualified Unicode.Internal.Char.Security.IntentionalConfusables as IC
+
+-- [TODO] @since
+-- | Identifier status
+--
+data IdentifierStatus
+    = Restricted
+    -- ^ /Restricted/ characters should be treated with caution when considering
+    -- possible use in identifiers, and should be disallowed unless there is
+    -- good reason to allow them in the environment in question.
+    | Allowed
+    -- ^ /Allowed/ characters are not typically used as is by implementations.
+    -- Instead, they are applied as filters to the set of supported characters.
+    deriving (Eq, Ord, Bounded, Enum, Show)
+
+-- [TODO] @since
+-- | Returns the 'IdentifierStatus' corresponding to a character.
+--
+{-# INLINE identifierStatus #-}
+identifierStatus :: Char -> IdentifierStatus
+identifierStatus c = if S.isAllowedIdentifier c
+    then Allowed
+    else Restricted
+
+-- [TODO] @since
+-- | Returns 'True' if the given character is 'Allowed' in an identifier.
+--
+{-# INLINE isAllowedIdentifier #-}
+isAllowedIdentifier :: Char -> Bool
+isAllowedIdentifier = S.isAllowedIdentifier
```

Comment:
We can rename this to `isIdentifierAllowed`.

Reply:
Not sure of this: on one hand `isAllowedIdentifier` sounds better English but suggests we check a whole identifier; on the other hand `isIdentifierAllowed` suggests a check of only a part of an identifier, but does not sound as good English. I suggest `isAllowedInIdentifier` which is explicit.

Reply:
Depends on how you read it, and where you put the punctuation. "is identifier allowed?" as a questions is correct English. "is allowed, identifier" is a bit odd whereas "is allowed-identifier" is correct because we are asking if it is an allowed-identifier. Anyway both are ok with me.